### PR TITLE
Protect iterators against concurrent modification

### PR DESCRIPTION
**Pull request by mejrs**
Fixes https://github.com/PyO3/pyo3/issues/2162

 - [x] Changelog
- [x] More tests?
- [x] Does PyList need to be protected? It seems fine as is.

### CODE REVIEW

File: src/types/dict.rs [hunk 1]
```diff
@@ -213,13 +213,13 @@ impl PyDict {
 
     /// Returns an iterator of `(key, value)` pairs in this dictionary.
     ///
-    /// Note that it's unsafe to use when the dictionary might be changed by
-    /// other code.
+    /// # Panics
+    ///
+    /// If PyO3 detects that the dictionary is mutated during iteration, it will panic.
+    /// It is allowed to modify the values of the keys as you iterate over the dictionary, but only
```
Comment (birkenfeld) [inline src/types/dict.rs:219]: I know "modify the values of the keys" is copied verbatim from the C-API docs, but I think it's quite confusing.

File: src/types/dict.rs [hunk 2]
```diff
@@ -228,51 +228,129 @@ impl PyDict {
...
+#[cfg(all(Py_3_8, not(PyPy), not(Py_LIMITED_API)))]
+mod impl_ {
+    use super::*;
+    use std::marker::PhantomData;
+
+    pub struct PyDictIterator<'py> {
+        di_dict: *mut ffi::PyDictObject,
+        di_pos: ffi::Py_ssize_t,
+        di_used: ffi::Py_ssize_t,
+        marker: PhantomData<&'py PyDict>,
```
Comment (davidhewitt) [inline src/types/dict.rs:240]: Is there a risk that the dict refcount can fall to 0 and can be accidentally released? Should this be `Py<PyDict>`?

File: src/types/dict.rs [hunk 3]
```diff
@@ -228,51 +228,129 @@ impl PyDict {
...
-                ))
-            } else {
-                None
+    impl<'a> std::iter::IntoIterator for &'a PyDict {
+        type Item = (&'a PyAny, &'a PyAny);
+        type IntoIter = PyDictIterator<'a>;
+
+        fn into_iter(self) -> Self::IntoIter {
+            unsafe {
+                let di_dict: *mut ffi::PyDictObject = self.as_ptr().cast();
```
Comment (davidhewitt) [inline src/types/dict.rs:294]: I've always wondered if there should be an `as_raw() -> *mut ffi::PyDictObject` method on `PyDict` (and similar for other types)? Available only on not (limited_api / PyPy).

File: src/types/dict.rs [hunk 4]
```diff
@@ -228,51 +228,129 @@ impl PyDict {
...
+        ///
+        /// # Panics
+        ///
+        /// If PyO3 detects that the dictionary is mutated during iteration, it will panic.
+        /// It is allowed to modify the values of the keys as you iterate over the dictionary, but only
+        /// so long as the set of keys does not change.
+        #[inline]
+        fn next(&mut self) -> Option<Self::Item> {
+            unsafe {
+                if self.di_used != (*(self.di_dict)).ma_used {
```
Comment (davidhewitt) [inline src/types/dict.rs:256]: What's the difference between `ma_used` and `PyDict_Size` used below?

File: src/types/dict.rs [hunk 5]
```diff
@@ -228,51 +228,129 @@ impl PyDict {
...
+                    self.di_used = -1;
+                    panic!("dictionary changed size during iteration");
```
Comment (davidhewitt) [inline src/types/dict.rs:258]: Panicking seems much nicer than returning results 👍

File: src/types/frozenset.rs [hunk 6]
```diff
@@ -0,0 +1,194 @@
...
+    PyFrozenSet,
+    ffi::PySetObject,
+    ffi::PyFrozenSet_Type,
+    #checkfunction=ffi::PyFrozenSet_Check
+);
+
+impl PyFrozenSet {
+    /// Creates a new frozenset.
+    ///
+    /// May panic when running out of memory.
```
Comment (davidhewitt) [inline src/types/frozenset.rs:25]: I know not from this PR, but is this actually true? It looks like returning `PyResult` means that we'll actually return `PyErr` containing a `MemoryError` if we run out of memory.

I suspect that actually the text may want to be more like `PySet` where we don't mention this and instead comment on the error return being caused by unhashable contents.

File: src/types/frozenset.rs [hunk 7]
```diff
@@ -0,0 +1,194 @@
+// Copyright (c) 2017-present PyO3 Project and Contributors
+//
+
```
Comment (davidhewitt) [inline src/types/frozenset.rs:3]: 👍 good idea to make a new file.

**Review by davidhewitt — COMMENTED** [review]
I love this! 🚀 

> Does PyList need to be protected? It seems fine as is.

Good question. I guess it's by luck that with the current implementation we re-fetch the `len()` on each call to `next()`? I guess checking for modifications would require that _and_ an extra check, so maybe it's ok as-is.

File: src/types/dict.rs (hunk 2 above)
Comment (mejrs) [reply src/types/dict.rs:240]: The iterator borrows from the Pydict, so this should be fine.

**Review by davidhewitt — APPROVED** [review]
Looks great, smart new implementation!

**Comment by davidhewitt** [comment]
I know I already approved... Looks like this needs a CHANGELOG and then good to merge?

**Comment by mejrs** [comment]
Yes, I've been busy for a bit. I'll do it soon :)

**Comment by mejrs** [comment]
That should be all 🎉

File: src/types/mod.rs [hunk 8]
```diff
@@ -32,6 +33,42 @@ pub use self::traceback::PyTraceback;
...
+/// })
+/// # }
+///  ```
+///
+/// If PyO3 detects that the collection is mutated during iteration, it will panic.
+///
+/// These iterators use Python's C-API directly. However in certain cases, like when compiling for
+/// the Limited API and PyPy, the underlying structures are opaque and that may not be possible.
+/// In these cases the iterators are implemented by forwarding to [`PyIterator`].
+pub mod iter {
```
Comment (davidhewitt) [inline src/types/mod.rs:66]: 👍 it was long overdue for these to be public!

**Review by davidhewitt — APPROVED** [review]
Looks great!

**Pull request merged**